src/isl/mod.rs

```diff
@@ -1,36 +1,76 @@
+//! Provides a way to construct ISL types/constraints programmatically into an internal representation.
+//!
+//! This module consists of three submodules that implement public facing API for constructing ISL types/constraint:
+//!
+//! * `isl_type` module represents a public facing schema type [IslType] which converts given ion content in the schema file
+//! into an internal representation of ISL types(not-yet-resolved types). It stores [IslConstraint]s defined within the given type.
+//! Also, this eventually gets converted to a solid [TypeRef](which contains a unique TypeId for the given type
+//! and a resolved type TypeDefinition) using [TypeStore].
+//!
+//! * `isl_constraint` module represents a public facing API for schema constraints [IslConstraint]
+//! which converts given ion content in the schema file into an internal representation of ISL constraint(not-yet-resolved constraints).
+//! This stores [IslTypeRef]s for the given ISL constraint.
+//!
+//! * `isl_type_reference` module provides an internal representation of a schema type reference.
+//! The type reference grammar is defined in the [Ion Schema Spec]: https://amzn.github.io/ion-schema/docs/spec.html#grammar
+//!
+//! ## Example usage of `isl` module to create an [IslType]:
+//! ```
+//! use ion_rs::IonType;
+//! use ion_schema_rust::isl::{isl_type::*, isl_constraint::*, isl_type_reference::*};
+//!
+//! fn main() {
+//!     // below code represents an ISL type:
+//!     // `type:: {
+//!     //      name:my_type_name,
+//!     //      type: int,
+//!     //      all_of: [
+//!     //          { type: bool }
+//!     //      ]
+//!     //  }
+//!     // `
+//!     let isl_type = IslType::named(
+//!         // represents the `name` of the defined type
+//!         "my_type_name".to_owned(),
+//!         vec![
+//!             // represents the `type: int` constraint
+//!             IslConstraint::type_constraint(
+//!                 IslTypeRef::core_isl(IonType::Integer)
+//!             ),
+//!             // represents `all_of` with anonymous type `{ type: bool }` constraint
+//!             IslConstraint::all_of(
+//!                 vec![
+//!                     IslTypeRef::anonymous(
+//!                         vec![
+//!                             IslConstraint::type_constraint(
+//!                                 IslTypeRef::core_isl(IonType::Boolean)
+//!                             )
+//!                         ]
+//!                     )
+//!                 ]
+//!             )
+//!         ]
+//!     );
+//!     assert_eq!(isl_type.constraints().len(), 2);
```

Comment:
TODO: We might consider actually creating a schema system and validating something trivial here unless that's too verbose.

Reply:
Maybe we can add an example for schema system within the schema system module and keep this as it is for it represents `isl` module example. Thoughts?

Reply:
My sense is that anyone reading the module documentation is hoping to figure out how to build a `Schema` from `Isl*` types. (Said another way: an end user would only build an `IslType` programmatically for one reason.)  I think this example should show building a `Schema` from an `IslType` (via the `SchemaSystem`) and the schema system module's example can show building a `Schema` from some an Ion/ISL document. The latter could also link to the former.

Reply:
Reference issue: https://github.com/amzn/ion-schema-rust/issues/33